addons/hr_payroll/models/hr_leave.py

```diff
@@ -0,0 +1,71 @@
+# -*- coding:utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+from dateutil.relativedelta import relativedelta
+from odoo import api, fields, models, _
+from odoo.addons.resource.models.resource import Intervals
+
+class HrLeaveType(models.Model):
+    _inherit = 'hr.leave.type'
+
+    benefit_type_id = fields.Many2one('hr.benefit.type', string='Benefit Type')
+
+
+class HrLeave(models.Model):
+    _inherit = 'hr.leave'
+
+    @api.multi
+    def copy_to_benefits(self):
+        for leave in self:
+            benefit_type = leave.holiday_status_id.benefit_type_id
+            self.env['hr.benefit'].safe_duplicate_create({
+                'name': "%s%s" % (benefit_type.name + ": " if benefit_type else "", leave.employee_id.name),
+                'date_start': leave.date_from,
+                'date_stop': leave.date_to,
+                'benefit_type_id': benefit_type.id,
+                'employee_id': leave.employee_id.id,
+                'leave_id': self.id,
+            })
+
+    @api.multi
+    def _cancel_benefit_conflict(self):
+        benefits = self.env['hr.benefit'].search([('leave_id', 'in', self.ids)])
+        if benefits:
+            self.copy_to_benefits()
+
+            # create new benefits where the leave does not cover the full benefit
+            benefits_intervals = Intervals(intervals=[(b.date_start, b.date_stop, b) for b in benefits])
+            leave_intervals = Intervals(intervals=[(l.date_from, l.date_to, l) for l in self])
+            remaining_benefits = benefits_intervals - leave_intervals
+
+            for interval in remaining_benefits:
+                benefit = interval[2]
+                leave = benefit.leave_id
+                benefit_type = benefit.benefit_type_id
+                employee = benefit.employee_id
+
+                benefit_start = interval[0] + relativedelta(seconds=1) if leave.date_to == interval[0] else interval[0]
+                benefit_stop = interval[1] - relativedelta(seconds=1) if leave.date_from == interval[1] else interval[1]
+
+                self.env['hr.benefit'].safe_duplicate_create({
+                    'name': "%s: %s" % (benefit_type.name, employee.name),
+                    'date_start': benefit_start,
+                    'date_stop': benefit_stop,
+                    'benefit_type_id': benefit_type.id,
+                    'employee_id': employee.id,
+                })
+            benefits.unlink()
+
+    @api.multi
+    def action_validate(self):
+        super(HrLeave, self).action_validate()
+        self._cancel_benefit_conflict()
```

Comment:
Actually, it does not add much protection to forbid the access of benefits for Holidays Officer/Manager/Team Leader:

benefits are only an interface to create `resource.calendar.[leaves | attendance]` when they are validated.
If we grant them access, the only additional info we would give are employees' work schedules, which is not private info, is it?
If a user from Holidays group tries to validate benefits, it would be ok for  `resource.calendar.leaves` but throw an AccessError because they lack access for `resource.calendar.attendance`.

To solve the issue above, I would give access rights to benefits for Holidays groups.
What do you think? Would you prefer not and use `sudo()` for holiday users in `action_validate()` and `action_refuse()` ?

Reply:
We should try to follow the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege). If a role does not *need* access to a table/record, then they should not have it, especially not _write_ access.
It seems risky to let "team leaders" mess up with actual hr.benefit entries, even if only by mistake (reminder: the access won't be restricted to just `action_validate()`, they will have full CRUD). These benefits records are meant to be used for payroll purposes, ultimately generating payslips, so they must be carefully controlled and verified by Payroll staff. In some companies it will be unthinkable to give this power to team leaders.
So yes, it seems you should probably sudo() just for those calls.